Zend/zend_operators.h

```diff
@@ -919,6 +919,63 @@ static zend_always_inline void zend_unwrap_reference(zval *op) /* {{{ */
 }
 /* }}} */
 
+static zend_always_inline bool zend_strnieq(const char *ptr1, const char *ptr2, size_t num)
+{
+	const char *end = ptr1 + num;
+	while (ptr1 < end) {
+		if (zend_tolower_ascii(*ptr1++) != zend_tolower_ascii(*ptr2++)) {
+			return 0;
+		}
+	}
+	return 1;
+}
+
+static zend_always_inline const char *
+zend_memnistr(const char *haystack, const char *needle, size_t needle_len, const char *end)
+{
+	ZEND_ASSERT(end >= haystack);
+
+	if (UNEXPECTED(needle_len == 0)) {
+		return haystack;
+	}
+
+	const char first_lower = zend_tolower_ascii(*needle);
+	const char first_upper = zend_toupper_ascii(*needle);
+	const char *p_lower = (const char *)memchr(haystack, first_lower, end - haystack);
+	const char *p_upper = (first_lower != first_upper) ? (const char *)memchr(haystack, first_upper, end - haystack) : NULL;
+	const char *p = !p_upper || (p_lower && p_lower < p_upper) ? p_lower : p_upper;
+
+	if (needle_len == 1) {
+		return p;
+	}
+
+	size_t off_s = (size_t)(end - p);
+	if (needle_len > off_s) {
+		return NULL;
+	}
+
+	const char needle_end_lower = zend_tolower_ascii(needle[needle_len - 1]);
+	const char needle_end_upper = zend_toupper_ascii(needle[needle_len - 1]);
+	end -= needle_len;
+
+	while (p && p <= end) {
```

Comment:
This `if` is bothering me :P If we reorganize variables: `if (needle_len > end - p)` -> `if (p > end - needle_len)` seems like `if` and `while` are redundant now. Even more: `p <= end` could be removed if `end -= needle_len` would be moved before the first memchr. Currently after the first iteration `p <= end` will always be true. Or i missing something?

Reply:
Of course this is relaying that `needle_len < haystack_len`, since negative `end - haystack` could be bad.

Reply:
Yeah, it's a micro-optimization to avoid the `zend_to(lower,upper)_ascii` calls that's probably not very impactful.

> Of course this is relaying that needle_len < haystack_len, since negative end - haystack could be bad.

As long as `p` isn't dereferenced there shouldn't be a problem. The check in the while would rule that out (since `p` will always be `>= haystack`. The `if` is also wrong actually, `p` might be `NULL`, we're relying on the fact that `needle_len` is smaller than `end` which is very very likely but still an incorrect assumption.

I'll remove the `if`. Thanks!

Reply:
> Yeah, it's a micro-optimization to avoid the `zend_to(lower,upper)_ascii` calls that's probably not very impactful.

You can still do that in earlier if: `if (needle_len == 1 || !p) {`